### PR TITLE
Assembly joints

### PR DESCRIPTION
**Pull request by PaddleStroke**
This is a basic implementation of a constraint system for the assembly wb.

The approach is the one that uses 'joint connectors'. It is a similar approach as A4 but the idea is to generate the placement based on the user selection. So for example you select the plane face4 of your body, the joint-connector placement is going to be the rotation of that plane, and the position of the face4 vertex that is the closer to your mouse. It is also similar to the approach of two well known commercial package.

The reason of this choice is that the solver developped by Dr Koh is following this logic to build assembly. Also this workflow is good and intuitive.

- 'Create Assembly' also create a 'joints' folder.
- Adds 'create joint xxx' commands. These commands prompt the user to select 2 features of 2 different parts/bodies.
- The command creates a 'joint' object that stores the selected objects and selected elements and generated placement in properties.

The joint-connectors are ready to be assembled. At this stage we now need to have the API of the solver ready and accessible to move forward and actually solve the assembly.

Note: 
- Commit 1 concerns core and has a PR already opened : #10573
- Issue https://github.com/FreeCAD/FreeCAD/issues/10467
Early feedback would be appreciated.

### CODE REVIEW

File: src/Mod/Assembly/JointObject.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    - If elt is a vertex, then vtx = elt And placement is vtx coordinates without rotation.
+    - if elt is an edge, then vtx = edge start/end vertex depending on which is closer. If elt is an arc or circle, vtx can also be the center. The rotation is the plane normal to the line positioned at vtx. Or for arcs/circle, the plane of the arc.
+    - if elt is a plane face, vtx is the face vertex (to the list of vertex we need to add arc/circle centers) the closer to the mouse. The placement is the plane rotation positioned at vtx
+    - if elt is a cylindrical face, vtx can also be the center of the arcs of the cylindrical face.
+    """
+
+    def findPlacement(obj, elt, vtx):
+        plc = obj.Placement
+
+    return plc
```
Comment (howie-j) [inline src/Mod/Assembly/JointObject.py:181]:
```suggestion
        return plc
```
return outside of method, gives error when loading workbench:
asm_error

Comment (PaddleStroke) [reply src/Mod/Assembly/JointObject.py]: Sorry about that typo !

**Comment by PaddleStroke** [comment]
This is now ready for review.

File: src/Mod/Assembly/CommandCreateJoint.py [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+
+    def GetResources(self):
+        tooltip = "<p>Create a revolute joint between parts of your assembly."
+        tooltip += "This will allow rotation along one axis.</p>"
+
+        return {
+            "Pixmap": "Assembly_CreateJointRevolute",
+            "MenuText": QT_TRANSLATE_NOOP("Assembly_CreateJointRevolute", "Create Revolute Joint"),
+            "Accel": "J,R",
+            "ToolTip": QT_TRANSLATE_NOOP("Assembly_CreateJointRevolute", tooltip),
```
Comment (chennes) [inline src/Mod/Assembly/CommandCreateJoint.py:85]: This won't work, you can't use a variable in QT_TRANSLATE_NOOP.

File: src/Mod/Assembly/JointObject.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+                self.joint.Object2, self.joint.Element2, self.joint.Vertex2
+            )
+        else:
+            self.joint.Object2 = None
+            self.joint.Element2 = ""
+            self.joint.Vertex2 = ""
+            self.joint.Placement2 = UtilsAssembly.activeAssembly().Placement
+
+    """
+    So here we want to find a placement that correspong to a local cordinate system that would be placed at the selected vertex.
```
Comment (chennes) [inline src/Mod/Assembly/JointObject.py:173]:
```suggestion
    So here we want to find a placement that corresponds to a local coordinate system that would be placed at the selected vertex.
```

File: src/Mod/Assembly/UtilsAssembly.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+
+def getElementName(full_name):
+    # full_name is "Assembly.Assembly1.Assembly2.Assembly3.Box.Edge16"
+    # We want either Edge16.
+    parts = full_name.split(".")
+
+    if len(parts) < 3:
+        # At minimum "Assembly.Box.edge16". It shouldn't be shorter
+        return ""
```
Comment (chennes) [inline src/Mod/Assembly/UtilsAssembly.py:92]: As opposed to raising an exception?

File: src/Mod/Assembly/UtilsAssembly.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    points = []
+    for vtx in vertexes:
+        points.append(vtx.Point)
+    return points
+
+
+def findClosestPointToMousePos(candidates_points, mousePos):
+    point_min_length = None
+    i = 0
+    for point in candidates_points:
```
Comment (chennes) [inline src/Mod/Assembly/UtilsAssembly.py:223]: I'd use `enumerate` here.

File: src/Mod/Assembly/UtilsAssembly.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+            point_min_length = length
+            closest_point_index = i
+        i += 1
+
+    return closest_point_index, point_min_length
+
+
+def findVertexNameInObject(vertex, obj):
+    i = 0
+    for vtx in obj.Shape.Vertexes:
```
Comment (chennes) [inline src/Mod/Assembly/UtilsAssembly.py:235]: Again, `enumerate` is more idiomatic.

**Comment by howie-j** [comment]
### Non-code feedback

Note: I've mostly tested the workflow of adding joints, since that is what this PR adresses.

I find it very intuitive and easy to add joints, really great work so far. I also like the agile approach of making small PR's, quickly merging, getting early feedback and solving problems continually.

#### :+1: :+1: :+1:
- The LCS indicator snaps to edges, vertexes, faces and holes mostly as expected, and the preview is awesome
- Nice to be able to edit joint type in task view, and later in property view
- Quick to hide LCS's/joints by clicking "Joints" folder and space bar
- Overall i'm very impressed by the user experience so far. Not only miles ahead of the other wb's, but also on par with commercial systems.

#### :-1: 
- Joint/LCS snap has room for improvement, for example snap to mid line and center of face
- The LCS's / joints are quite big and intrusive, and should perhaps be hidden automatically after a joint is created
- Enter should confirm joint, to be consistent with task panels (PD Pad, Pocket etc)
- Double clicking on a joint in tree view should launch the joint task panel for easy editing
- No way of selecting and removing a joint connector in the task panel (by right clicking in the list for example)
- Selecting a joint connector in the task panel list and pressing delete will delete the last feature of the part, not remove the joint connector
- If you delete the "Joints" folder/group, you get an error when attempting to create a new joint. The create joint command should check if a Joints folder exists, and create a new one if not
- (Very personal opinion) The shortcuts follows the horrible sketcher model of multiple keys, when Assembly currently only has 10 commands. In my opinion it's better to only create default shortcuts for the most used commands, and let the user themself create shortcuts for less used commands. Much better to have 5 or 10 single key shortcuts that the user might remember, than 30 that the user wont.. :)

File: src/Mod/Assembly/CommandCreateJoint.py [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+    def deactivated(self):
+        Gui.Selection.removeSelectionGate()
+        Gui.Selection.removeObserver(self)
+        Gui.Selection.setSelectionStyle(0)
+        Gui.Selection.clearSelection()
+        self.view.removeEventCallback("SoLocation2Event", self.callbackMove)
+
+    def createJointObject(self):
+        type_index = self.form.jointType.currentIndex()
+
```
Comment (howie-j) [inline src/Mod/Assembly/CommandCreateJoint.py:360]: Check if group "Joints" exist i assembly, and create if if not

**Comment by PaddleStroke** [comment]
> 1 - [ ] The LCS's / joints are quite big and intrusive, and should perhaps be hidden automatically after a joint is created
> 2 - [ ] Enter should confirm the joint, to be consistent with task panels (PD Pad, Pocket etc)
> 3 - [ ] Double clicking on a joint in tree view should launch the joint task panel for easy editing
> 4 - [ ] No way of selecting and removing a joint connector in the task panel (by right clicking in the list for example)
> 5 - [ ] Selecting a joint connector in the task panel list and pressing delete will delete the last feature of the part, not remove the joint connector
> 6 - [x] If you delete the "Joints" folder/group, you get an error when attempting to create a new joint. The create joint command should check if a Joints folder exists, and create a new one if not
> 7 - [ ] Joint/LCS snap has room for improvement, for example snap to mid line and center of face (nitpicking, not important atm)
> 8 - [x] (Very personal opinion) The shortcuts follows the horrible sketcher model of multiple keys, when Assembly currently only has 10 commands. In my opinion it's better to only create default shortcuts for the most used commands, and let the user themself create shortcuts for less used commands. Much better to have 5 or 10 single key shortcuts that the user might remember, than 30 that the user wont.. :)

- 1, 3, 4, 7 : Yes I agree. But each of them need more work. I was thinking that these could be subsequent issues / PR, as to involve more people in the development.
- 5 : Yes this is an undefined behavior. I think as delete is undefined in the list, the event goes back to the parent, until it reaches the comboview, and there it does the tree delete action. I would say it goes with 4.
- 6 and 8 done

- 2 : When you want to catch key press events in the 3D view, you need to use addEventCallback: 

```
        self.callbackKey = self.view.addEventCallback("SoKeyboardEvent", self.KeyboardEvent)
        self.view.removeEventCallback("SoLocation2Event", self.callbackKey)

    def KeyboardEvent(self, info):
        if info["State"] == "UP" and info["Key"] == "ESCAPE":
            # Create a qt key release event for the Escape key
            key_release_event = QtGui.QKeyEvent(
                QtGui.QKeyEvent.KeyPress, QtCore.Qt.Key_Escape, QtCore.Qt.NoModifier
            )
            # Send the key release event to the taskbox
            QtGui.QApplication.sendEvent(self.form, key_release_event)

        if info["State"] == "UP" and info["Key"] == "RETURN":
            if self.accept():
                Gui.Control.closeDialog()
```
The question is how to close the task. I found the 2 options above, first make a QKeyEvent to act as if user pressed esc in the taskbox and send it to the taskbox. Second call reject or accept, and if true then closeDialog.
@chennes could you please advise what is the proper way to dismiss a task when the 3d view has the focus?

Besides I found a weird bug with this. In ways it works ONCE but if you start the tool a second time again and try to do it again (give focus to view and click enter or escape), then it crash.

**Comment by qewer33** [comment]
I also just gave the branch a test and I really like it! The LCS selection is incredibly intuitive. I agree with the points mentioned by howie-j but I'd like to add a few little things:
- I realized that some of the joint icons have improper padding making them appear small on the toolbar, that's my bad. I can fix it in a PR later when this is merged or you can also address it in this PR if you want @PaddleStroke
- It would be great to have alignment options for the part LCSs in the task view (I'm thinking like two groups, "Edit First Part LCS" and "Edit Second Part LCS" with options to rotate, flip, offset etc.). This might be better if it's implemented after this base PR is merged though, but would love to see that
- When will the Ondsel Solver be added so we can test the actual solver joint functionality?
- howie-j mentioned how LCSs would be better hidden, to add on that, it would be great if the LCSs and their attached vertices/features can be highlighted when the joint is selected in the tree view

**Comment by PaddleStroke** [comment]
@qewer33 
- Please go ahead and make a PR, as it's separated there won't be conflicts.
- Alignment options to rotate / reverse the placement is something we need to implement indeed.
- I'm not sure when the API will be ready. Hopefully soon !
- Yes that's needed too

**Comment by chennes** [comment]
> could you please advise what is the proper way to dismiss a task when the 3d view has the focus?

I think that in this case both of your suggestions are reasonable. I normally don't care for generating keypress events, but since yours is really _passing along_ a keypress event it seems like an acceptable path to me. I'd personally probably take the other route, and call accept or reject, but I don't feel strongly about it.

**Comment by PaddleStroke** [comment]
Ok then I'll go with 

``` 
  if self.accept():
      Gui.Control.closeDialog()
```

Are you guys able to reproduce the crash I'm having when closing the tool with Escape for the second time ? If so do you have any logs on your side that could hint at what is wrong? I have no message on my end.

**Comment by howie-j** [comment]
Yes, esc crashes freecad.
Steps to reproduce:
- Switch to assembly
- Create new assembly
- Click create joint (any)
- Click in the 3D view
- Double click escape key, freecad crashes

[FreeCAD.log](https://github.com/FreeCAD/FreeCAD/files/12640586/FreeCAD.log)

**Comment by chennes** [comment]
That looks like it's trying to send an event to a Widget that has been deleted.

**Comment by jpg87Fr** [comment]
Hello everyone.
I would like to know what to do to test the progress of the new Assembly WB.

**Comment by qewer33** [comment]
> Hello everyone. I would like to know what to do to test the progress of the new Assembly WB.

You can clone the branch of this PR and build it but it currently doesn't have the solver integrated, you may want to test it after that's done.

**Comment by chennes** [comment]
> I would like to know what to do to test the progress of the new Assembly WB.

Right now there's not much to see -- are you interested in testing the un-merged PR code, or the code that's already been integrated into the master branch?

**Comment by PaddleStroke** [comment]
I just forced push with : 
- Fixing the crash. For reference, it was a copy paste error in the removeEventCallback of the keyboard event:
```
        self.view.removeEventCallback("SoLocation2Event", self.callbackMove)
        self.view.removeEventCallback("SoLocation2Event", self.callbackKey)
```
instead of 
```
        self.view.removeEventCallback("SoLocation2Event", self.callbackMove)
        self.view.removeEventCallback("SoKeyboardEvent", self.callbackKey)
```
- Improve the tooltips of the joint commands.

**Comment by PaddleStroke** [comment]
A summary of what needs to be done in subsequent PRs:

- [ ] JCS are ugly.
- [ ] JCS should hide when the joint is not in edit.
- [ ] When Joint is selected in the tree, the JCS and selected feature should be highlighted.
- [ ] Joints should be editable. Double clicking should open the Task
- [ ] Delete a joint connector in the taskbox list should remove it from selection.
- [ ] JCS should also snap to line middle and face center of gravity.
- [ ] Add buttons to rotate / reverse JCS

If any of you want to pick up one of these let me know !
@qewer33 maybe you'd be interested to make a better design for the JCS? You can just make a mockup and I'll try to make it in coin. Or if you want to do it in coin directly you are more than welcome 😄

**Comment by chennes** [comment]
OK, this looks ready to me then, I'm going to merge it.

**Comment by qewer33** [comment]
> If any of you want to pick up one of these let me know ! @qewer33 maybe you'd be interested to make a better design for the JCS? You can just make a mockup and I'll try to make it in coin. Or if you want to do it in coin directly you are more than welcome 😄

Personally, I find the current LCS design usable. Just checked OnShape and Fusion360 and they seem to have a very similar graphic. They will definitely look better a little smaller though, can you test that?

**Comment by qewer33** [comment]
Ok I couldn't hold myself and made a mockup anyway @PaddleStroke :sweat_smile: 

It's similar to the current one, just a little shorter and thicker lines with a translucent plane to show the local base XY plane.

image

Here's what OnShape has for reference but I don't think it's that great.

image

**Comment by jpg87Fr** [comment]
_> are you interested in testing the un-merged PR code, or the code that's already been integrated into the master branch?_
I don't know anything about code. I will wait until new elements are present in the Weekly version.

**Pull request merged**